### PR TITLE
[SPARK-13302][PYSPARK][TESTS] Move the temp file creation and cleanup outside of the doctests

### PR DESCRIPTION
**Pull request by holdenk**
Some of the new doctests in ml/clustering.py have a lot of setup code, move the setup code to the general test init to keep the doctest more example-style looking.
In part this is a follow up to https://github.com/apache/spark/pull/10999 
Note that the same pattern is followed in regression & recommendation - might as well clean up all three at the same time.

### CODE REVIEW

**Comment by holdenk** [comment]
cc @mengxr who was interested in this cleanup

File: python/pyspark/ml/clustering.py [hunk 1]
```diff
@@ -310,7 +303,17 @@ def _create_model(self, java_model):
...
     globs['sqlContext'] = sqlContext
-    (failure_count, test_count) = doctest.testmod(globs=globs, optionflags=doctest.ELLIPSIS)
-    sc.stop()
+    import tempfile
+    temp_path = tempfile.mkdtemp()
+    globs['temp_path'] = temp_path
+    try:
+        (failure_count, test_count) = doctest.testmod(globs=globs, optionflags=doctest.ELLIPSIS)
+        sc.stop()
+    finally:
```
Comment (yanboliang) [inline python/pyspark/ml/clustering.py]: I think the ```try ... finally``` is not necessary because it does not handle any exception.

**Comment by yanboliang** [comment]
@holdenk I like this cleanup, thanks for the effort.

File: python/pyspark/ml/clustering.py (hunk 1 above)
Comment (holdenk) [inline python/pyspark/ml/clustering.py]: So finally is still useful even if we don't explicitly catch/handle any exceptions - are you saying the sc.stop and doctest will never throw any exceptions?

Comment (yanboliang) [inline python/pyspark/ml/clustering.py]: Sorry for misunderstand, I think your are right.

**Comment by yanboliang** [comment]
LGTM

**Comment by holdenk** [comment]
great :)

**Comment by holdenk** [comment]
maybe @mengxr if you've got a chance to take a look at this?

**Comment by holdenk** [comment]
or @davies or @jkbradley maybe?

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**